frontend/src/utils/handleCaptionsSignal/handleCaptionsSignal.ts

```diff
@@ -0,0 +1,156 @@
+  captionsActiveCountRef: RefObject<number>;
+  currentRoomNameRef: RefObject<string | null>;
+  vonageVideoClient: VonageVideoClient | null;
+  setCaptionsEnabled: Dispatch<SetStateAction<boolean>>;
+};
+
+/**
+ * Handles the captions signal received from the session.
+ * This function manages enabling, disabling, joining, and leaving captions.
+ * It also handles the functionality of notifying other participants about the current captions ID and the participant count.
+ * @param {CaptionsHandleType} props - Dependencies needed for the handler
+ * @property {SignalEvent} event - The signal event containing the data.
+ * @property {RefObject<string | null>} currentCaptionsIdRef - Reference to the current captions ID
+ * @property {RefObject<number>} captionsActiveCountRef - Reference to the count of active participants using captions
+ * @property {RefObject<string | null>} currentRoomNameRef - Reference to the current room name
+ * @property {RefObject | null} vonageVideoClient - The Vonage Video client instance
+ */
+const handleCaptionsSignal = ({
+  event,
+  currentCaptionsIdRef,
+  captionsActiveCountRef,
+  currentRoomNameRef,
+  vonageVideoClient,
+  setCaptionsEnabled,
+}: CaptionsHandleType) => {
+  try {
+    const parsedData: CaptionsSignalDataType = JSON.parse(event.data as string);
+    const { action, captionsId, currentCount } = parsedData;
+
+    switch (action) {
+      case 'enable':
+        if (captionsId) {
+          currentCaptionsIdRef.current = captionsId;
+          setCaptionsEnabled(true);
+          vonageVideoClient?.signal({
+            type: 'captions',
+            data: JSON.stringify({
+              action: 'update-current-user-count',
+              currentCount: captionsActiveCountRef.current + 1,
+            }),
+          });
+        }
+        break;
+
+      case 'join':
+        setCaptionsEnabled(true);
+        captionsActiveCountRef.current += 1;
+        break;
+
+      case 'update-current-user-count':
+        if (currentCount) {
+          captionsActiveCountRef.current = currentCount;
+        }
+        break;
+
+      case 'leave': {
+        const newCount = Math.max(0, captionsActiveCountRef.current - 1);
+        setCaptionsEnabled(false);
+
+        // If there are no other participants using captions, we disable them for the whole session.
+        // This is to ensure that captions are only disabled when there are other participants using them.
+        if (newCount === 0 && currentCaptionsIdRef.current && currentRoomNameRef.current) {
+          disableCaptions(currentRoomNameRef.current, currentCaptionsIdRef.current)
+            .then(() => {
+              currentCaptionsIdRef.current = null;
+
+              vonageVideoClient?.signal({
+                type: 'captions',
+                data: JSON.stringify({ action: 'disable' }),
+              });
+            })
+            .catch((err) => console.error('Error disabling captions:', err));
```

Comment:
Could we handle this through the backend instead? There's likely to be race conditions when everyone leaves the call at the same time. Alternatively, do we need to disable captions if there's nobody in the call? My understanding of pricing is that only published minutes are counted. The expiration for the captions seems sufficient.

Reply:
This actually won't get triggered if you simply leave the session. Maybe I could rephrase it a little better but the phrase "if there are no other participants using captions" implies that the person currently disabling the captions is the last one to do so by clicking the "disable captions" button but they are still staying in the session